subsys/sensing/dispatch.c

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright (c) 2022 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <zephyr/sys/__assert.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/rtio/rtio.h>
+#include <zephyr/sensing/sensing_sensor.h>
+#include "sensor_mgmt.h"
+
+LOG_MODULE_DECLARE(sensing, CONFIG_SENSING_LOG_LEVEL);
+
+/* check whether it is right time for client to consume this sample */
+static inline bool sensor_test_consume_time(struct sensing_sensor *sensor,
+				     struct sensing_connection *conn,
+				     uint64_t cur_time)
+{
+	LOG_DBG("sensor:%s next_consume_time:%lld cur_time:%lld",
+			sensor->dev->name, conn->next_consume_time, cur_time);
+
+	return conn->next_consume_time <= cur_time;
+}
+
+static void update_client_consume_time(struct sensing_sensor *sensor,
+				       struct sensing_connection *conn)
+{
+	uint32_t interval = conn->interval;
+
+	if (conn->next_consume_time == 0) {
+		conn->next_consume_time = get_us();
+	}
+
+	conn->next_consume_time += interval;
+}
+
+/* send data to clients based on interval and sensitivity */
+static int send_data_to_clients(struct sensing_sensor *sensor,
+				void *data)
+{
+	struct sensing_sensor *client;
+	struct sensing_connection *conn;
+
+	for_each_client_conn(sensor, conn) {
+		client = conn->sink;
+		LOG_DBG("sensor:%s send data to client:%p", conn->source->dev->name, conn);
+
+		if (!is_client_request_data(conn)) {
+			continue;
+		}
+
+		/* sensor_test_consume_time(), check whether time is ready or not:
+		 * true: it's time for client consuming the data
+		 * false: client time not arrived yet, not consume the data
+		 */
+		if (!sensor_test_consume_time(sensor, conn, get_us())) {
```

Comment:
This seems like it would drop data if the consumer is not yet expecting it based on get_us(). I'd note that ticks are imprecise and this could fail to pass the branch in cases that we might otherwise expect.

Reply:
I think we discussed this in the AWG before we created the sensor WG. We shouldn't do any filtering in the subsystem. It's better to present the client with ALL the data that's available and let them choose a decimation strategy.